### ISSUE TITLE
Test-Time Adaptation via Self-Training with Nearest Neighbor Information

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#25]
### Paper Link
https://openreview.net/forum?id=EzLtB4M1SbM
### Code Link
https://github.com/mingukjang/TAST
## Motivation
Self-training methods use predictions from the classifier as pseudo labels for the test data and fine-tune the classifier to make it fit to the pseudo labels. These methods have a limitation that the fine-tuned classifier can overfit to the inaccurate pseudo labels, resulting in confirmation bias.

## Methodology

<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/229768632-61fb4113-732c-425c-91eb-45a5021934f1.png">

<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/229774225-934651fc-3d46-47c8-9d7a-7b8f54693dac.png">

### Prototype-based Classification in Test-Time Adaptation

- Obtain prototype representations, which represent each class in the embedding space, and then predict the label of an input as the class of the nearest prototype.
- $\mathbb{S}_t=\left\lbrace\mathbb{S}_t^1, \mathbb{S}_t^2, \cdots, \mathbb{S}_t^K\right\rbrace$ : a set of test samples until time t, K is the total number of classes. 
- At time t, the support set is updated as (Eq. 1) if the prediction entropy is less than the threshold:

$$
\begin{equation}
\mathbb{S}_t^k= \begin{cases}\mathbb{S}_{t-1}^k \cup\left\lbrace\frac{f_\theta\left(x_t\right)}{\left\|f_\theta\left(x_t\right)\right\|}\right\rbrace & \text { if arg } \max _c p_c=k, \mathbb{S}_{t-1}^k & \text { otherwise },\end{cases}
\end{equation}
$$

- Formally, the prototype $\mu_k$ for class k is computed as 

$$\mu_k=\frac{1}{\left|\mathbb{S}_t^k\right|} \sum_{z \in \mathbb{S}_t^k} z$$

- The prediction for an input $x_t$ is made by comparing the distances between the embedding of $x_t$ and the prototypes:

$$
\hat{y}=\arg \min _c d\left(f_\theta\left(x_t\right), \mu_c\right)
$$

### Adaptation Module
- Randomly initialized new classifier
- During the test time, we train the adaptation module and predict the label of the test data
- Train the adaptation modules independently and predict the label of the test data using the average predicted class distribution from the adaptation modules (ensemble)

### Self-training with NN 
1. Find $N_s$ nearby support examples (Eq. 2) $\mathcal{N}(x ; \mathbb{S}):=\left\lbrace z \in \mathbb{S} \mid d\left(f_\theta(x), z\right) \leq \beta_x\right\rbrace$

2. Prototype-based class prediction: $p_i^{\text {proto }}(k \mid z):=\frac{\exp \left(-d\left(h_{\phi_i}(z), \mu_{i, k}\right) / \tau\right)}{\sum_c \exp \left(-d\left(h_{\phi_i}(z), \mu_{i, c}\right) / \tau\right)}$

3. Generate a pseudo label (one hot) by aggregating prototype-based class prediction of the nearby support examples in $\mathcal{N}(x ; \mathbb{S})$ as (Eq. 4):

$$
\hat{p}_i^{\text {TAST }}(k \mid x):=\frac{1}{N_s} \sum_{z \in \mathcal{N}(x ; \mathbb{S})} \mathbb{1}\left[\underset{c}{\arg \max } p_i^{\text {proto }}(c \mid z)=k\right]
$$

4. Fine-tune the adaptation modules: $\mathcal{L}^{\mathrm{TAST}}\left(\phi_i\right)=\frac{1}{\left|D^{\text {test }}\right|} \sum_{x \in D^{\text {test }}} \operatorname{CE}\left(\hat{p}_i^{\mathrm{TAST}}(\cdot \mid x), p_i^{\text {proto }}(\cdot \mid x)\right)$
5. Obtain new prediction (Eq. 6):

$$
p_i^{\text {proto }}(k \mid x):=\frac{\exp \left(-d\left(h_{\phi_i}\left(f_\theta(x)\right), \mu_{i, k}\right) / \tau\right)}{\sum_c \exp \left(-d\left(h_{\phi_i}\left(f_\theta(x)\right), \mu_{i, c}\right) / \tau\right)}, k=1,2, \ldots, K
$$

6. Iterate the pseudo labeling and fine-tuning (step 3 to 5) processes for $T$ steps per batch.
7. Final prediction is averaged by from the adaptation modules: 

$$
\begin{aligned}
p_i^{\text {TAST }}(k \mid x) & :=\frac{1}{N_s} \sum_{z \in \mathcal{N}(x ; \mathbb{S})} p_i^{\mathrm{proto}}(k \mid z) \\
\hat{y}^{\mathrm{TAST}} & =\underset{c}{\arg \max } p^{\mathrm{TAST}}(c \mid x)=\underset{c}{\arg \max } \frac{1}{N_e} \sum_{i=1}^{N_e} p_i^{\text {TAST }}(c \mid x)
\end{aligned}
$$

- TAST-BN fine-tunes the BN layers instead of adaptation modules.

## Implementation 

- Set the batch size to 128

## Experiments 

<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/229957298-71d8defa-9b3d-446b-b466-a6fbb8b4403e.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229957587-72148d7a-273a-44a2-9c3d-1ec6a189c1d8.png">

## Notes
- This method utilizes Prototype + NN + learnable ensemble modules to improve the quality of pseudo labels during test-time.  How to improve the quality of pseudo labels really matter. 
- This method requires multiple training for each batch and may fail if batch size is small.
